### PR TITLE
feat: add intermediate result YAML saving

### LINKED ISSUE #158: Provide procedure/interface to interrupt fit and pick up latest fit result

### PR DESCRIPTION
**Pull request by sebastianJaeger**
I added a functionality that regularly saves intermediate fit results, so not all of the progress is lost upon abortion of the fit or a death of the process.

Closes #158

### CODE REVIEW

File: src/tensorwaves/optimizer/minuit.py [hunk 1]
```diff
@@ -69,3 +81,18 @@ def __func(pars: list) -> float:
...
         results["time"] = end_time - start_time
         return results
+
+
+def dump_parameters(parameters: dict, outputfile: str) -> None:
+    output = {
+        "Parameters": {
+            name: float(value) for name, value in parameters.items()
+        }
+    }
```
Comment (spflueger) [inline src/tensorwaves/optimizer/minuit.py:91]: I would probably just dump the dictionary as it is, without wrapping it in the `"Parameters"` (That's a bit of a preference though). I probably would put parameters in the automated temporary filename. Maybe @redeboer has an opinion about this.

File: src/tensorwaves/optimizer/minuit.py [hunk 2]
```diff
@@ -69,3 +81,18 @@ def __func(pars: list) -> float:
...
+    with open(outputfile, "w") as yaml_file:
+        yaml.dump(
+            output,
+            yaml_file,
+            sort_keys=False,
+            default_flow_style=False,
+        )
```
Comment (spflueger) [inline src/tensorwaves/optimizer/minuit.py:98]: Just an idea (that feature does not have to also be implemented here): We could think about a feature that appends new Parameters to have really a path of how the Optimizer traveled.

**Review by spflueger — COMMENTED** [review]
@sebastianJaeger Thanks for the feature! I added a few comments.

One general thing is, that it is rather hard to test this functionality nicely. But that's more of a design issue, and I would propose to implement this functionality in a decorator (the decorator can be tested with a mock quite nicely). 

If you do now have fun doing this, I can understand. But this would be a perfect time to test our new branching model in action. The idea is to push this code into a new "feature" branch. In this branch the "code quality" are not as strict as in the master branch, but allow quick additions of new features. Then those branches can be refactored and merge with the master later on. @redeboer can tell a bit more about the details

File: src/tensorwaves/optimizer/minuit.py [hunk 3]
```diff
@@ -14,13 +17,17 @@ class Minuit2(Optimizer):
...
+        self, filename: Optional[str] = None, n_iterations: int = 10
+    ) -> None:
+        self.__filename = filename
+        self.__n_iterations = n_iterations
 
     def optimize(self, estimator: Estimator, initial_parameters: dict) -> dict:
         parameters = initial_parameters
 
         function_calls = 0
+        progress_bar = tqdm()
```
[the remark this answers is not in the document]
Comment (sebastianJaeger) [reply src/tensorwaves/optimizer/minuit.py:30]: It does not get to know when its finished, but at least you get to see how fast you run, sicne it shows the number of iterations and the time per iteration, so you have a rough guess how quickly the computation goes. In my opinion thats still better than no current status output at all.

File: src/tensorwaves/optimizer/minuit.py (hunk 2 above)
Comment (sebastianJaeger) [reply src/tensorwaves/optimizer/minuit.py:98]: That might me usefull, but I dont think that should be enabled by default, since it probably complicates the original task this was implemented for.

**Comment by redeboer** [comment]
Here are a few thoughts:
- We should try to follow the interfaces of [keras optimizers](https://www.tensorflow.org/api_docs/python/tf/keras/optimizers/Optimizer)
- If we write to a format that can be understood by [tensorboard](https://www.tensorflow.org/tensorboard/scalars_and_keras), we may be able to use `tensorboard` to monitor fit progress.
- For that we need to write to different sequential log files, see also https://github.com/ComPWA/tensorwaves/pull/154#discussion_r521906639

> This would be a perfect time to test our new branching model in action. The idea is to push this code into a new "feature" branch. In this branch the "code quality" are not as strict as in the master branch, but allow quick additions of new features. Then those branches can be refactored and merge with the master later on. @redeboer can tell a bit more about the details

Since this concerns quite a few issues, I created a branch [99-export-fit](https://github.com/ComPWA/tensorwaves/tree/epic/99-export-fit) that is to address #99.

File: src/tensorwaves/optimizer/minuit.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (redeboer) [reply src/tensorwaves/optimizer/minuit.py:30]: Update: #157 implements a `ProgressBar` callback that prints the current estimator value in addition to the time progressed. We can also print other stats that way.

File: src/tensorwaves/optimizer/minuit.py (hunk 2 above)
Comment (redeboer) [reply src/tensorwaves/optimizer/minuit.py:98]: This could be handled by creating a new `Callback` implementation (#164) and registering values in, say, a `pandas.DataFrame`.

Side note: it would be nice to have the fit running in the background (`asyncio`?), so that you can just `Minuit.optimize` in a notebook and monitor (dynamically -- not by continuously reading back a file) the fit parameters in that `DataFrame`.

**Comment by redeboer** [comment]
I updated this PR a bit with the new callback syntax (#164). Now it should also close #158, as illustrated in the notebook.

@sebastianJaeger, have a look here for the new implementation
https://github.com/ComPWA/tensorwaves/pull/154/commits/2d621a560ad7b6d13fbf33e4c2c61632c63b7a54?file-filters%5B%5D=.py#diff-9f533f5231a0e5fcd84806765d6bbf0282451172465fee983ac5df640128cceeR34-R77

File: src/tensorwaves/optimizer/minuit.py (hunk 1 above)
Comment (redeboer) [reply src/tensorwaves/optimizer/minuit.py]: I still wrapped the parameters in a subsection, because there's also a `"Time"` section and `"EstimatorValue"`.

**Review by redeboer — APPROVED** [review]

File: src/tensorwaves/optimizer/callbacks.py [hunk 4]
```diff
@@ -30,6 +31,50 @@ def finalize(self) -> None:
...
+
+        .. code-block:: bash
+
+            tensorboard --logdir logs
+        """
+        self.__iteration = 0
+        self.__step_size = step_size
+        self.__stream = open(filename, "w")
+
+    def __call__(self, parameters: dict, estimator_value: float) -> None:
```
Comment (spflueger) [inline src/tensorwaves/optimizer/callbacks.py:53]: The Callbacks are actually called on every function evaluation... Unfortunately we don't have access to the iterations (at least when using minuit).

File: src/tensorwaves/optimizer/callbacks.py [hunk 5]
```diff
@@ -30,6 +31,50 @@ def finalize(self) -> None:
...
+        self.__iteration += 1
+        if self.__iteration % self.__step_size != 0:
+            return
+        output_dict = {
+            "Time": datetime.now(),
+            "Iteration": self.__iteration,
+            "EstimatorValue": float(estimator_value),
```
Comment (spflueger) [inline src/tensorwaves/optimizer/callbacks.py:60]: It would be good if we would be able to pass the estimator label through as well. Maybe through the init?

File: src/tensorwaves/optimizer/callbacks.py [hunk 6]
```diff
@@ -30,6 +31,50 @@ def finalize(self) -> None:
...
+            "Parameters": {
+                name: float(value) for name, value in parameters.items()
+            },
+        }
+        _empty_file(self.__stream)
```
Comment (spflueger) [inline src/tensorwaves/optimizer/callbacks.py:65]: we dont want to kept appending?

**Review by spflueger — APPROVED** [review]
All in all looks good. Just some small remarks that are not urgent (since its only going into the epic branch).

File: src/tensorwaves/optimizer/callbacks.py (hunk 6 above)
Comment (redeboer) [reply src/tensorwaves/optimizer/callbacks.py:65]: No, that would conflict with YAML syntax

File: src/tensorwaves/optimizer/callbacks.py (hunk 4 above)
Comment (redeboer) [reply src/tensorwaves/optimizer/callbacks.py:53]: Yeah, that was one of the reasons for posting #165

File: src/tensorwaves/optimizer/callbacks.py (hunk 5 above)
Comment (redeboer) [reply src/tensorwaves/optimizer/callbacks.py]: Have a look 171b6e5 (it can be reverted)

File: src/tensorwaves/optimizer/callbacks.py (hunk 6 above)
Comment (spflueger) [reply src/tensorwaves/optimizer/callbacks.py:74]: Not if everything is wrapped into an array right? But ok that can be a future improvement

Comment (redeboer) [reply src/tensorwaves/optimizer/callbacks.py:74]: Yeah but then YAML is perhaps not the appropriate format (rather csv or so)

Comment (spflueger) [reply src/tensorwaves/optimizer/callbacks.py:74]: Just another idea. Maybe Yaml is also not the best choice here. And the structure. If we would append there is like a 70% overhead in terms of storage, since the parameter names would be written to disk every time. Some table structure with a header would be more useful.

File: src/tensorwaves/optimizer/callbacks.py (hunk 5 above)
Comment (spflueger) [reply src/tensorwaves/optimizer/callbacks.py]: Looks ok. But would be a bit inconvenient for appending  see https://github.com/ComPWA/tensorwaves/pull/154#discussion_r525941074

File: src/tensorwaves/optimizer/callbacks.py (hunk 4 above)
Comment (spflueger) [reply src/tensorwaves/optimizer/callbacks.py:59]: I meant more that `iterations` should be renamed to `function_calls`. As iteration is the incorrect word here

**Comment by spflueger** [comment]
I added some more ideas. But depending on how fast this functionality is needed it can be merged to the epic branch and improved there later on. Hence i approved the PR
